### PR TITLE
Chore/consistent type hints

### LINKED ISSUE #94: Use type hints consistently
As part of that: where a variable could be either `pd.Series` or `pd.DataFrame`, do sth like the following:

```
DataFrameOrSeries = Union[pd.DataFrame, pd.Series]

def compute_ma(
    data: DataFrameOrSeries, fun: Callable, spans: List[int], plot: bool = True
) -> pd.DataFrame:
    # rest of function
```

While at it, fix pylint issues as well.

### LINKED ISSUE #118: Update documentation
Some of the new classes, e.g. market and asset are currently not in the documentation
When you are bored... Redo documentation to include proper support for type hints (after type hints are consistently implemented and tested with mypy).

### PR DESCRIPTION
**Pull request by fmilthaler**
Using type hints throughout FinQuant and introducing type checks in functions/methods that raise appropriate error messages if an argument is of an unexpected type.
This commit also introduces custom data types, e.g. `FLOAT` as `Union[np.floating, float]` and more to simplify type hints. Moreover, type validation is now done with a new module/function `type_validation` that also simplifies type validations.
Finally, the documentation for all modules in `finquant` has been updated.
This closes  #94 and also closes #118.

### CODE REVIEW

**Comment by fmilthaler** [comment]
@PietropaoloFrisoni This is finally ready to be merged. 

### Important Note
Please, do not review this thoroughly. I only assigned this PR to you, so that you can have a quick look at some of the changes I point out to you. Please ignore everything else, it is just too much to review. :)

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -35,9 +35,28 @@ jobs:
...
+        continue-on-error: true
 
-      - name: Analysing the code with pylint
-        if: ${{ matrix.python-version }} == '3.10'
+      - name: Pylint analysis
+        id: pylint_analysis
         run: |
-          pylint --fail-under=9 $(git ls-files '*.py')
\ No newline at end of file
+          python -m pylint --fail-under=10 $(git ls-files '*.py')
```
Comment (fmilthaler) [inline .github/workflows/pytest.yml:45]: Any PR CI runs are now going to fail if Pylint finds anything it is not happy about.

File: .github/workflows/pytest.yml [hunk 2]
```diff
@@ -35,9 +35,28 @@ jobs:
...
+        continue-on-error: true
+
+      - name: mypy analysis
+        id: mypy_analysis
+        run: |
+          python -m mypy *.py finquant
```
Comment (fmilthaler) [inline .github/workflows/pytest.yml:51]: Any PR CI runs are now going to fail if mypy finds anything it is not happy about. Making sure that type hints are implemented correctly.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -35,6 +35,24 @@ your branch name should be something like bugfix/print-statement-portfolio-prope
 For the automated versioning to work, the branch name is required to start with `bugfix/` or one of the other
 above mentioned patterns.
 
+### Custom data types
```
Comment (fmilthaler) [inline CONTRIBUTING.md:38]: Please read through the changes in this file and let me know if it is clear.

File: finquant/data_types.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+from typing import Any, KeysView, List, TypeVar, Union
+
+import numpy as np
+import pandas as pd
+
+# Generic List Element Type
+ELEMENT_TYPE = TypeVar("ELEMENT_TYPE")
+
+# Type Aliases:
+ARRAY_OR_LIST = Union[np.ndarray[ELEMENT_TYPE, Any], List[ELEMENT_TYPE]]
```
Comment (fmilthaler) [inline finquant/data_types.py:60]: Have a quick look at the custom defined data types here (up to `STRING_OR_DATETIME`.

File: finquant/minimise_fun.py [hunk 5]
```diff
@@ -3,37 +3,65 @@
 """
 
 
+from finquant.data_types import ARRAY_OR_DATAFRAME, ARRAY_OR_SERIES, FLOAT, NUMERIC
 from finquant.quants import annualised_portfolio_quantities
+from finquant.type_utilities import type_validation
 
 
-def portfolio_volatility(weights, mean_returns, cov_matrix):
+def portfolio_volatility(
```
Comment (fmilthaler) [inline finquant/minimise_fun.py:11]: This function definition is a good example of the custom type hints being used

File: finquant/minimise_fun.py [hunk 6]
```diff
@@ -3,37 +3,65 @@
...
+    :type mean_returns: :py:data:`~.finquant.data_types.ARRAY_OR_SERIES`
+
+    :param cov_matrix: Covariance matrix of returns
+    :type cov_matrix: :py:data:`~.finquant.data_types.ARRAY_OR_DATAFRAME`
+
+    :rtype: :py:data:`~.finquant.data_types.FLOAT`
+    :return: Annualised volatility
     """
+    # Type validations:
+    type_validation(weights=weights, means=mean_returns, cov_matrix=cov_matrix)
```
Comment (fmilthaler) [inline finquant/minimise_fun.py:31]: In this line we call the new function `type_validation` that takes care of validating the types of given arguments. Once this PR is merged, please follow this practice as well.

File: finquant/minimise_fun.py [hunk 7]
```diff
@@ -3,37 +3,65 @@
...
 from finquant.quants import annualised_portfolio_quantities
+from finquant.type_utilities import type_validation
 
 
-def portfolio_volatility(weights, mean_returns, cov_matrix):
+def portfolio_volatility(
+    weights: ARRAY_OR_SERIES[FLOAT],
+    mean_returns: ARRAY_OR_SERIES[FLOAT],
+    cov_matrix: ARRAY_OR_DATAFRAME[FLOAT],
+) -> FLOAT:
```
Comment (fmilthaler) [inline finquant/minimise_fun.py:15]: This function definition is a good example of the custom type hints being used

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -35,6 +35,24 @@ your branch name should be something like bugfix/print-statement-portfolio-prope
...
+
+To accommodate and simplify this, custom data types are defined in the module `finquant.data_types`.
+Please familiarize yourself with those and add more if your code requires them.
+
+### Data type validation
+[FinQuant provides a module/function for type validation](https://finquant.readthedocs.io/en/latest/developers.html#type-validation),
+which is used throughout the code base for type validation purposes. Said function simplifies checking an argument
+against its expected type and reduces the amount of copy-pasted `if` and `raise` statements.
+You can check out the source code in `finquant.type_utilities`.
+
```
Comment (fmilthaler) [inline CONTRIBUTING.md:55]: Please read through the changes in this file and let me know if it is clear.

File: finquant/data_types.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+# To support Dict listkeys:
+LIST_DICT_KEYS = Union[ARRAY_OR_LIST[ELEMENT_TYPE], KeysView[ELEMENT_TYPE]]
+
+# Numeric types
+FLOAT = Union[np.floating, float]
+INT = Union[np.integer, int]
+NUMERIC = Union[INT, FLOAT]
+
+# String/Datetime types
+STRING_OR_DATETIME = Union[str, datetime]
```
Comment (fmilthaler) [inline finquant/data_types.py:74]: Have a quick look at the custom defined data types here.

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (fmilthaler) [reply .github/workflows/pytest.yml:45]: This is quite strict, meaning the code quality according to Pylint needs to get a score of 10/10. If you say it is a bit too strict, I'm happy for this to go down to 9/10. Your decision :)

**Comment by PietropaoloFrisoni** [comment]
Congrats man! This is quite a careful and thorough work. it certainly brings `FinQuant` to the next level : )

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (PietropaoloFrisoni) [reply .github/workflows/pytest.yml:45]: For me, 10 is fine (indeed very instructive).

**Review by PietropaoloFrisoni — APPROVED** [review]

**Review by PietropaoloFrisoni — APPROVED** [review]

**Pull request merged**